Alternator: add slow query logging

psarna opened the pull request:
This series adds slow query logging capability to alternator. Queries which last longer than the specified threshold are logged in `system_traces.node_slow_log` and traced.

In order to be better prepared for https://github.com/scylladb/scylla/issues/2572, this series also expands the tracing API to allow custom key-value params and adds a custom `alternator_op` parameter to the slow node log. This information can also be deduced from the tracing session id by consulting the system_traces.events table, but https://github.com/scylladb/scylla/issues/2572 's assumption is that this tracing might not always be available in the future.

This series comes with a simple test case which checks if operation logs indeed end up in `system_traces.node_slow_log`.

Tests:
unit(dev, alternator pytest)
manual: verified that no operations are logged if slow query logging is disabled; verified that operations that take less time than the threshold are not logged; verified with test_batch.py::test_batch_write_item_large that a large-enough operation is indeed logged and traced.

Fixes #8292

Example trace:

```cql
cqlsh> select parameters, duration from system_traces.node_slow_log where start_time=b7a44589-8711-11eb-8053-14c6c5faf955;

 parameters                                                                                  | duration
---------------------------------------------------------------------------------------------+----------
 {'alternator_op': 'DeleteTable', 'query': '{"TableName": "alternator_Test_1615979572905"}'} |    75732
```

psarna commented:
I updated the description to add a paragraph about testing.

nyh commented:
I wonder if we need this new alternator_op thing - couldn't we just inserted into the "query", which can contain, say, the operation followed by the JSON parameters? This will require a string copy, but does it matter for tracing which slow anyway?

nyh commented on alternator/server.cc:
```diff
@@ -319,10 +319,11 @@ future<> server::verify_signature(const request& req, const chunked_content& con
 }
 
-static tracing::trace_state_ptr create_tracing_session() {
+static tracing::trace_state_ptr create_tracing_session(tracing::tracing& tracing_instance) {
```
The fact you're now passing this tracing_instance parameter is just an optimization, to avoid calling get_local_tracing_instance() twice, right? This change wasn't really necessary here, right?
(I'm not against it, just trying to understand the motivation).

psarna replied:
It's for two things:
 - reducing the line length
 - removing the dependency on a global variable; it's still there one level up, but eventually we might want to remove that one as well

nyh commented on test/alternator/test_tracing.py:
```diff
@@ -70,6 +71,45 @@ def with_tracing(dynamodb):
         pytest.skip('Failed to verify tracing disabled')
 
+# Similarly to the fixture above, slow query logging is enabled only for the run of the
+# test function.
```
Would be nice to write here what you this fixture sets the "slow query" threshold to be. I see below "1". Is that 1us? 1ms? Or what?

psarna replied:
Oh, it was supposed to be 0 actually, but 1 is equally effective. It's in microseconds, I'll add a small comment explaining what happens.

nyh commented on test/alternator/test_tracing.py:
```diff
@@ -212,3 +252,24 @@ def test_tracing_all(with_tracing, test_table_s_isolation_always, dynamodb):
+    delay = 0.2
+    while delay < 30:
+        results = full_scan(slow_query_table, ConsistentRead=False)
+        put_item_found = any("PutItem" in result['parameters'] and p in result['parameters'] for result in results)
```
Looks good. I see now why you wanted to have this "PutItem" somewhere in the trace. I agree it's useful to have it but not it is not stricty necessary for the test - what I did in previous test is to use a different random string as the key for each operation, so just searching for this random key was enough to guarantee this operation was logged.

psarna commented:
> I wonder if we need this new alternator_op thing - couldn't we just inserted into the "query", which can contain, say, the operation followed by the JSON parameters? This will require a string copy, but does it matter for tracing which slow anyway?

I considered it, but I'd prefer not to make tracing even slower. It will also involve more allocations, and my assumption is that people turn on slow query tracing if their queries get slow, which might also suggest that the cluster is under stress. I'm waiting for Vlad's opinion about the additional param, but, to be honest, just ignoring the alternator op would be as good a solution - a bit less comfortable for the users, but it would still be trivial to distinguish what kind of operation it was just by looking at the parameters. So if `alternator_op` is for any reason not acceptable, I'd go with just not putting this information in `system_traces.node_slow_log` at all.

psarna commented:
v2:
 - changed the threshold from 1us to 0us (just because that was my initial idea)
 - added a comment about it in the slow query logging test fixture

nyh commented:
> > I wonder if we need this new alternator_op thing - couldn't we just inserted into the "query", which can contain, say, the operation followed by the JSON parameters? This will require a string copy, but does it matter for tracing which slow anyway?
> 
> I considered it, but I'd prefer not to make tracing even slower. It will also involve more allocations, and my assumption is that people turn on slow query tracing if their queries get slow, which might also suggest that the cluster is under stress. I'm waiting for Vlad's opinion about the additional param, but, to be honest, just ignoring the alternator op would be as good a solution - a bit less comfortable for the users, but it would still be trivial to distinguish what kind of operation it was just by looking at the parameters. So if `alternator_op` is for any reason not acceptable, I'd go with just not putting this information in `system_traces.node_slow_log` at all.

I think it's ok as you did it. If Vlad or anyone else objects later, you can always send another patch to change the tracing. I'm building and testing your patches now and plan to commit them.

vladzcloudius reviewed (changes requested):
This series does A LOT more than needed in order to enable Slow Query Logging with Alternator, which should have been a matter of looking at a single bit and do a few trivial manipulations if it's set.
@psarna 
I suggest you split this series into at least two:
1) Enable Slow Query Logging (should be a trivial patch).
2) Enhance Tracing (in general) of Alternator.

The following relates to the (2).

'query': '{"TableName": "alternator_Test_1615979572905"}' doesn't look very helpful.

`query` is (usually) supposed to tell us what is the operation that is being performed.

In your example I think a more readable option would be something like this (if you like JSON a lot):

```
'query': '{"op" : "DeleteTable", "TableName" : "alternator_Test_1615979572905"}'
```

I don't mind adding more generic values to `parameters` if needed however here you'd rather enrich your 'query' key.

BTW how come a table name ended in the 'query' key - there is a separate API for it: `tracing::add_table_name`.

psarna commented:
@vladzcloudius the `query` contains exactly what the alternator user sent to us - and in case of `DeleteTable` op, it also includes the table name as one of the fields. JSON is only here because I chose not to touch these potentially large strings in order to avoid redundant copying. It also has a slight advantage that this `query` can be copy-pasted and rerun without changes, which won't be the case if I modify the JSON by adding some additional fields.

psarna commented:
And as for `add_table_name`, it's also used in alternator tracing, but I suppose it will only be visible in the actual traces (from system_traces.events), not in the node_slow_log table.

nyh commented:
> This series does A LOT more than needed in order to enable Slow Query Logging with Alternator, which should have been a matter of looking at a single bit and do a few trivial manipulations if it's set.
> @psarna
> I suggest you split this series into at least two:

I already committed this series. The series was already split into three patches, and I don't really mind that it contains other small cleanups around the relevant code, in addition to just enabling the slow query logging.

> 
>     1. Enable Slow Query Logging (should be a trivial patch).
> 
>     2. Enhance Tracing (in general) of Alternator.

I agree that this would have been ideal, but I don't think this reorganization is important enough to revert the commit I already did.
I think the content of the code is, at the end, more important than the "purity" of the patches. 

> 'query': '{"TableName": "alternator_Test_1615979572905"}' doesn't look very helpful.
> 
> `query` is (usually) supposed to tell us what is the operation that is being performed.

I tend to agree, and also suggested that it can be a concatenation of the operation name and the query body (these parameters). But I also don't see any real harm in them being separate. It's all just text that users look at, anyway.
 
> In your example I think a more readable option would be something like this (if you like JSON a lot):
> 
> ```
> 'query': '{"op" : "DeleteTable", "TableName" : "alternator_Test_1615979572905"}'

I suggested even simpler "op {...}". 
 
> BTW how come a table name ended in the 'query' key - there is a separate API for it: `tracing::add_table_name`.

Piotr had a good point that all these different tracing attributes should really be different per protocol. Maybe CQL has a notion of a single "table name" per request, but the DynamoDB API doesn't, and there are requests which involve multiple tables. The table name is just a part of the request parameters, and some requests (e.g., BatchGetItem) may have multiple different tables mentioned.

vladzcloudius commented:
> > This series does A LOT more than needed in order to enable Slow Query Logging with Alternator, which should have been a matter of looking at a single bit and do a few trivial manipulations if it's set.
> > @psarna
> > I suggest you split this series into at least two:
> 
> I already committed this series. 

Hmmm... What's the point of requesting a review (from me) if you are committing regardless... :/

> The series was already split into three patches, and I don't really mind that it contains other small cleanups around the relevant code, in addition to just enabling the slow query logging.
> 
> > ```
> > 1. Enable Slow Query Logging (should be a trivial patch).
> > 
> > 2. Enhance Tracing (in general) of Alternator.
> > ```
> 
> I agree that this would have been ideal, but I don't think this reorganization is important enough to revert the commit I already did.
> I think the content of the code is, at the end, more important than the "purity" of the patches.

It's not only about the purity - it's more about the (1) being trivial and should go in easily.
While (2) requires some discussion.
For instance, your assumption that "tracing is slow anyway, so we can sustain as many extra string copies in its context as needed" is simply wrong in a general case. In particular in the context of #2572.

Slow Query Tracing "Light" (which is the only Slow Query Logging mode Cassandra has) should be as lightweight as possible because it's meant to be always enabled.

Therefore extra allocations, copies, formatting of strings should be avoided as much as possible in this mode. And in this mode we process everything that goes into `node_slow_log` and `sessions` tables.

> 
> > 'query': '{"TableName": "alternator_Test_1615979572905"}' doesn't look very helpful.
> > `query` is (usually) supposed to tell us what is the operation that is being performed.
> 
> I tend to agree, and also suggested that it can be a concatenation of the operation name and the query body (these parameters). But I also don't see any real harm in them being separate. It's all just text that users look at, anyway.
> 
> > In your example I think a more readable option would be something like this (if you like JSON a lot):
> > ```
> > 'query': '{"op" : "DeleteTable", "TableName" : "alternator_Test_1615979572905"}'
> > ```
> 
> I suggested even simpler "op {...}".
> 
> > BTW how come a table name ended in the 'query' key - there is a separate API for it: `tracing::add_table_name`.
> 
> Piotr had a good point that all these different tracing attributes should really be different per protocol. Maybe CQL has a notion of a single "table name" per request, but the DynamoDB API doesn't, and there are requests which involve multiple tables. The table name is just a part of the request parameters, and some requests (e.g., BatchGetItem) may have multiple different tables mentioned.

CQL has BATCH statements too. So, multi-table per single request is already supported (from day 0 Tracing has been introduced).

nyh commented:
> Hmmm... What's the point of requesting a review (from me) if you are committing regardless... :/

It's not contradicting. @psarna can still act on your review and submit followup patches. The only thing he cannot change after I committed it is the split of the series into separate patches. But he can change everything in the implementation based on your review, in the next patch.

> For instance, your assumption that "tracing is slow anyway, so we can sustain as many extra string copies in its context as needed" is simply wrong in a general case. In particular in the context of #2572.

I think you misunderstood @psarna 's explanation. I think he agreed with you exactly: The main reason why he didn't want to combine the operation name with the query parameters into one big "query" string is that he wanted to avoid an extra allocation. It sounds like you agree with his decision to keep them separate....

> > > BTW how come a table name ended in the 'query' key - there is a separate API for it: `tracing::add_table_name`.
> CQL has BATCH statements too. So, multi-table per single request is already supported (from day 0 Tracing has been introduced).

I'm not a tracing expert by any means, so please explain to me how tracing::add_table_name() works if there are multiple tables involved in the single batch query.
In any case, I'm not sure how this suggestion is related to this patch. Can you give an example where Alternator does the wrong thing by not using add_table_name() in some place?

Please note that the fact that the table name is part of the JSON-syntax query is a fact of life in the DynamoDB API. We're not going to change it, or remove it from the json. So maybe we can add a add_table_name() call (I think Piotr said we already have it?) but we can't and don't want to remove the table name from the query.
